src/general/world_effects/patch_events/GlobalGlaciationEvent.cs

```diff
@@ -0,0 +1,218 @@
+
+    public GlobalGlaciationEvent(GameWorld targetWorld, long randomSeed)
+    {
+        this.targetWorld = targetWorld;
+        random = new XoShiRo256starstar(randomSeed);
+    }
+
+    [JsonConstructor]
+    public GlobalGlaciationEvent(GameWorld targetWorld, XoShiRo256starstar random)
+    {
+        this.targetWorld = targetWorld;
+        this.random = random;
+    }
+
+    public void OnRegisterToWorld() { }
+
+    public void OnTimePassed(double elapsed, double totalTimePassed)
+    {
+        if (generationsLeft > 0)
+            generationsLeft -= 1;
+
+        if (generationsLeft > 0)
+            MarkPatches(totalTimePassed);
+
+        if (hasEventAlreadyHappened)
+            return;
+
+        if (generationsLeft == -1)
+            TryToTriggerEvent(totalTimePassed);
+        else if (generationsLeft == 0)
+            FinishEvent();
+    }
+
+    private void MarkPatches(double totalTimePassed)
+    {
+        foreach (var patch in targetWorld.Map.Patches.Values)
+        {
+            if (Constants.SurfaceBiomes.Contains(patch.BiomeType))
+            {
+                patch.AddPatchEventRecord(WorldEffectVisuals.GlaciationEvent, totalTimePassed);
+            }
+        }
+    }
+
+    private void TryToTriggerEvent(double totalTimePassed)
+    {
+        if (!AreConditionsMet())
+            return;
+
+        generationsLeft = random.Next(2, 6);
+
+        foreach (var (index, patch) in targetWorld.Map.Patches)
+        {
+            if (Constants.SurfaceBiomes.Contains(patch.BiomeType))
+            {
+                ChangePatchProperties(index, patch, totalTimePassed);
+            }
+        }
+    }
+
+    private bool AreConditionsMet()
+    {
+        var numberOfSurfacePatches = 0;
+        var patchesExceedingOxygenLevel = 0;
+        foreach (var patch in targetWorld.Map.Patches.Values)
+        {
+            if (!Constants.SurfaceBiomes.Contains(patch.BiomeType))
+                continue;
+
+            var oxygenLevel = patch.Biome.ChangeableCompounds[Compound.Oxygen];
+            if (oxygenLevel.Ambient >= Constants.OXYGEN_THRESHOLD)
+                patchesExceedingOxygenLevel += 1;
+
+            numberOfSurfacePatches += 1;
+        }
+
+        // Just prevent dividing by zero, but that shouldn't be possible anyway
+        numberOfSurfacePatches = numberOfSurfacePatches == 0 ? 1 : numberOfSurfacePatches;
+
+        return (float)patchesExceedingOxygenLevel / numberOfSurfacePatches >= Constants.OXYGEN_PATCHES_THRESHOLD
+            && random.Next(100) >= Constants.GLOBAL_GLACIATION_CHANCE;
+    }
+
+    private void ChangePatchProperties(int index, Patch patch, double totalTimePassed)
+    {
+        AdjustBackground(index, patch);
+        AdjustEnvironment(index, patch);
+        AddIceChunks(patch);
+        LogEvent(patch, totalTimePassed);
+    }
+
+    private void AdjustBackground(int index, Patch patch)
+    {
+        previousBackground.Add(index, patch.BiomeTemplate.Background);
```

Comment:
I think this needs to be refactored. All of the changed patch properties should be put into the patch snapshot system. I think it is the wrong approach to try to keep track of what every event has changed independently and try to then resolve the mess of 3 different events trying to restore a patch back to the way it should be.